src/smd-core.c

```diff
@@ -1000,7 +1006,10 @@ static int smd_attr_copy_val_to_internal(char *out, smd_dtype_t *t, smd_dtype_t
         case (SMD_TYPE_INT32): {
           int32_t ov = *(int32_t*)val;
           // check the range
-          if(ov < FLT_MIN || ov > FLT_MAX){
+          if(ov > 0 && (ov < FLT_MIN || ov > FLT_MAX)){
+            return 1;
+          }
+          if(ov < 0 && (ov > -FLT_MIN || ov < -FLT_MAX)){
```

Comment:
The `int32_t` to `float` conversion is the first one that may loose precision, but this check definitely won't catch it. The loss is like converting 123456789 to 1.234568e08: You cannot represent the number correctly, but it's definitely smaller than the largest representable number (9.999999e99).

I would resort to the following check instead: `if(ov != (int32_t)(float)ov) {...` This performs the conversion, converts back, and checks whether the original value comes out of the conversion. This catches exactly the cases where the integer value cannot be represented exactly.

Again, this is a systematic problem, and I won't point out every occurrence of it.

Reply:
We are checking the lost of precision on smd_attr_copy_val_to_external. See if that's enough.

Reply:
case (SMD_TYPE_INT32): {
         case (SMD_TYPE_FLOAT): {
            float ov = (float)(*p);
            // check if accuracy is precise enough
            if((int32_t) ov != *p){
              return 1;
            }
            *(float *)out = ov;
            return 0;
          }

Reply:
Ah, I see. Well, if checking precision is not the job of this function, there shouldn't be a check for range at all imho. The ranges of the `float` types are so vastly larger than what you can hope to get with integer types, that it hardly makes sense to check them.